lib/ts/redwood.ts

```diff
@@ -0,0 +1,169 @@
+/* Copyright (c) 2021, VRAI Labs and/or its affiliates. All rights reserved.
+ *
+ * This software is licensed under the Apache License, Version 2.0 (the
+ * "License") as published by the Apache Software Foundation.
+ *
+ * You may not use this file except in compliance with the License. You may
+ * obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+import Session from "./recipe/session/sessionRecipe";
+import { getSession, Error as STError } from "./recipe/session";
+
+import { ServerResponse } from "http";
+
+class CustomResponse extends ServerResponse {
+    chunks: Buffer[];
+
+    body: string | undefined;
+
+    write_original: Function;
+
+    end_original: Function;
+
+    write: any;
+
+    constructor({ options }: any) {
+        super({ options } as any);
+        this.chunks = [];
+        this.write_original = this.write;
+        this.end_original = this.end;
+        this.end = this._end_;
+        this.write = this._write_;
+        this.body = undefined;
+    }
+
+    _write_ = (chunk: any) => {
+        this.chunks.push(Buffer.from(chunk));
+        this.write_original(chunk);
+    };
+
+    _end_ = (chunk: any) => {
+        if (chunk) {
+            this.chunks.push(Buffer.from(chunk));
+        }
+
+        this.body = Buffer.concat(this.chunks).toString("utf8");
+        this.end_original(chunk);
+    };
+
+    json = (chunk: any) => {
+        this.setHeader("Content-Type", "application/json");
+        this.send(JSON.stringify(chunk));
+    };
+
+    set = (field: string, val: string | number | string[]) => {
+        let value = Array.isArray(val) ? val.map(String) : String(val);
+        this.setHeader(field, value);
+        return this;
+    };
+
+    get = (field: string) => {
+        return this.getHeader(field);
+    };
+
+    send = (chunk: any) => {
+        // populate Content-Length
+        let len = 0;
+        if (chunk !== undefined) {
+            if (Buffer.isBuffer(chunk)) {
+                len = chunk.length;
+            } else {
+                len = Buffer.from(chunk).length;
+            }
+
+            this.set("Content-Length", len);
+        }
+
+        // strip irrelevant headers
+        if (204 === this.statusCode || 304 === this.statusCode) {
```

Comment:
Why specifically these status codes only?

Reply:
Took this part from express response. Should I remove this?

Reply:
I don't know. See if it makes sense.. understand the code.. don't just blindly copy paste please.

Reply:
Response code 304 (Not Modified) indicates that the requested resource has not been modified since the previous transmission.
Response code 204 (No Content) success status indicates that a request has succeeded, but that the client doesn't need to navigate away from its current page.
For both these response codes, response is expected to have no content. So the above change just makes sure that any content related headers are stripped off.